cdap-watchdog/src/main/java/io/cdap/cdap/logging/ErrorLogsClassifier.java

```diff
@@ -219,6 +249,50 @@ private void populateResponse(IThrowableProxy throwableProxy,
     }
   }
 
+  @Nullable
+  private ErrorClassificationResponse findMatchingRule(CloseableIterator<LogEvent> logIter) {
+    while (logIter.hasNext()) {
+      ILoggingEvent logEvent = logIter.next().getLoggingEvent();
+      IThrowableProxy throwableProxy = logEvent.getThrowableProxy();
+      while (throwableProxy != null) {
+        for (ErrorClassificationRule rule : ruleList) {
+          if (throwableProxy.getClassName().equals(rule.getExceptionClassName())
+              && codeMethodMatched(rule.getCodeMethodRegex(),
+              throwableProxy.getStackTraceElementProxyArray())) {
+            LOG.debug("Rule {} matched", rule);
+            String errorMessage = throwableProxy.getMessage();
+            ErrorCategory errorCategory = new ErrorCategory(ErrorCategoryEnum.OTHERS, rule.getId());
+            return new ErrorClassificationResponse.Builder()
+                .setErrorCategory(errorCategory.getErrorCategory())
+                .setErrorType(rule.getErrorType().name())
+                .setErrorMessage(errorMessage)
+                .setThrowableClassName(throwableProxy.getClassName())
+                .build();
+          }
+          throwableProxy = throwableProxy.getCause();
+        }
+      }
+    }
+    return null;
+  }
+
+  private boolean codeMethodMatched(@Nullable String codeMethodRegex,
+      StackTraceElementProxy[] stackTraceElementProxies) {
+    if (Strings.isNullOrEmpty(codeMethodRegex)) {
+      return true;
+    }
+
+    Pattern pattern = Pattern.compile(codeMethodRegex);
+
+    for (StackTraceElementProxy stackTraceElementProxy : stackTraceElementProxies) {
```

Comment:
nit: can be also expressed as `return stackTraceElementProxies.stream().anyMatch(stackTraceElementProxy -> pattern.matcher(stackTraceElementProxy.getSTEAsString()).find());`

Reply:
done.